core/src/main/java/org/elasticsearch/cluster/service/InternalClusterService.java

```diff
@@ -179,6 +150,11 @@ public void setClusterStatePublisher(BiConsumer<ClusterChangedEvent, Discovery.A
         clusterStatePublisher = publisher;
     }
 
+    public void setLocalNode(DiscoveryNode localNode) {
+        DiscoveryNodes.Builder nodeBuilder = DiscoveryNodes.builder(clusterState.nodes()).put(localNode).localNodeId(localNode.id());
```

Comment:
will add the synchronization. Will also add an assertion that it is set once. Last, the assertion you requested is part of the doStart  method, I believe...